### ISSUE TITLE
Catch twitterbot on shared RECAP PDFs and insert twitter card data

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#863]
So, when people share PDFs on twitter, we don't have any card data associated with them. This is a missed opportunity.

To address this, we can tweak the code here:

https://github.com/freelawproject/courtlistener/blob/481e4f92e9a267200f67200169bcdcc7e6a90efa/cl/simple_pages/views.py#L336-L341

To watch for the Twitterbot user agent. When it's detected, we provide a useful HTML response page with twitter/facebook card information. When it's not, we serve up the PDF. 

I guess the next question is, what info do we want in the card. This is what we show on our HTML pages for documents:

```
{% block title %}{{ title|safe|striptags }} – CourtListener.com{% endblock %}
{% block og_title %}{{ title|safe|striptags }} – CourtListener.com{% endblock %}
{% block description %}{{ title|safe|striptags }} — Brought to you by the RECAP Initiative and Free Law Project, a non-profit dedicated to creating high quality open legal information.{% endblock %}
{% block og_description %}{{ title|safe|striptags }} — Brought to you by the
  RECAP Initiative and Free Law Project, a non-profit dedicated to creating
  high quality open legal information.{% endblock %}
```

And the `title` variable is defined as:

```
    title = '%sDocument #%s%s in %s' % (
        '%s &ndash; ' % item.description if item.description else '',
        item.document_number,
        ', Attachment #%s' % item.attachment_number if
        item.document_type == RECAPDocument.ATTACHMENT else '',
        best_case_name(item.docket_entry.docket),
    )
```

So the result is something like:

```
  <meta property="og:type" content="website"/>
  <meta property="og:title" content="Complaint &ndash; Document #1 in NATIONAL VETERANS LEGAL SERVICES PROGRAM v. United States – CourtListener.com"/>
  <meta property="og:description"
        content="Complaint &ndash; Document #1 in NATIONAL VETERANS LEGAL SERVICES PROGRAM v. United States — Brought to you by the
  RECAP Initiative and Free Law Project, a non-profit dedicated to creating
  high quality open legal information.">
  <meta property="og:url" content="https://www.courtlistener.com/docket/4214664/1/national-veterans-legal-services-program-v-united-states/"/>
  <meta property="og:site_name" content="CourtListener"/>
  <meta property="og:image"
        content="https://www.courtlistener.com/static/png/og-image-300x300.png"/>
  <meta property="og:image:type" content="image/png"/>
  <meta property="og:image:width" content="300"/>
  <meta property="og:image:height" content="300"/>
```

**Comment by mlissner** [comment]
We could probably improve this to show better values for `item.description` in the meta description and `og:description` fields. Right now we show the PACER short description (`Complaint`), but we could probably show the long description when we have that.

**Comment by mlissner** [comment]
The real fun here will be creating thumbnails of the PDFs, which seems like a noble pursuit, so that you can have a snapshot, a lá @big_cases bot.

**Comment by mlissner** [comment]
Hey, look, I did finish the code I was working on for this:

https://github.com/freelawproject/courtlistener/blob/9ee76e42ab4ea54edc9a8e35a9c78f4ccc40f731/cl/people_db/tasks.py#L10-L40

That's from the financial disclosure project, but it'll be trivial to generalize it. AWESOME. I think the performance should be there to do this on the fly so we don't have to do it for every PDF we get.

**Comment by mlissner** [comment]
This is pretty much done. I expect some trouble during deployment because there are a lot of weird file system things going on:

 - New directory for the thumbnails
 - Apache gets new alias
 - Can't really test this until it's live

But those shouldn't be a big deal really. 

I implemented this three times, each time refactoring to decrease complexity. Ugh:

1. First, I just copied the recap document template and tweaked it however I needed to make for a good twitter template.

2. Second, I realized there was too much overlap between Twitter's special template and the regular one, so I used includes and blocks to simplify the Twitter template.

3. Finally I realized there's no need for the special twitter template at all and that I can just use our NORMAL HTML page, but add the new Twitter stuff to it. 

Each refactoring wasn't too bad, but it was kind of dumb. Anyway, the implementation is very simple now. Whenever twitter comes crawling, instead of serving the PDF directly, we serve the regular recap document HTML page, which has the Twitter card info, and now has an embedded PDF. Simple.

---

Addressing the thoughts about what to put in the title and the description:

 - I upgraded the title a bit to simplify it, but not a ton. I played with adding the date and considered adding the author. Adding the date just...looked cluttered. Adding the author isn't something we have the data for. This is easy to further tweak if we want.

 - For the description, I changed it to show the full docket entry description if we have it. And if not, to show just the document short description. I think that'll work OK.

Tests are running now. If they pass, I'll deploy.
